yt/yt/experiments/public/fio_plugin/plugin.cpp

```diff
@@ -0,0 +1,821 @@
+
+#include <yt/yt/core/actions/future.h>
+
+#include <yt/yt/core/concurrency/async_stream.h>
+
+#include <yt/yt/library/auth/auth.h>
+
+#include <yt/yt/library/program/helpers.h>
+#include <yt/yt/library/program/program.h>
+#include <yt/yt/library/program/config.h>
+
+#include <library/cpp/yson/node/node_io.h>
+
+//#include <yt/yt/client/cache/rpc.h>
+
+#include <yt/yt/client/api/client.h>
+#include <yt/yt/client/api/config.h>
+#include <yt/yt/client/api/options.h>
+#include <yt/yt/client/api/file_reader.h>
+#include <yt/yt/client/api/file_writer.h>
+
+#include <yt/yt/client/chunk_client/helpers.h>
+
+#include <yt/yt/client/security_client/public.h>
+
+#include <yt/yt/client/api/rpc_proxy/config.h>
+#include <yt/yt/client/api/rpc_proxy/connection.h>
+
+#include <yt/yt/ytlib/api/native/config.h>
+#include <yt/yt/ytlib/api/native/connection.h>
+#include <yt/yt/ytlib/api/native/client.h>
+
+#include <yt/yt/ytlib/object_client/public.h>
+
+#include <yt/yt/ytlib/chunk_client/chunk_reader_host.h>
+#include <yt/yt/ytlib/chunk_client/replication_reader.h>
+#include <yt/yt/ytlib/chunk_client/chunk_meta_extensions.h>
+#include <yt/yt/ytlib/chunk_client/helpers.h>
+
+#include <yt/yt/ytlib/file_client/chunk_meta_extensions.h>
+
+#include <yt/yt/core/logging/log.h>
+
+//#include <yt/cpp/mapreduce/interface/logging/logger.h>
+
+#include <library/cpp/yt/memory/new.h>
+
+#include <yt/yt/core/ytree/yson_struct.h>
+
+// #include <yt/yt/core/logging/log.h>
+
+#include <util/system/env.h>
+#include <util/system/getpid.h>
+
+namespace NYT::NFio {
+
+using namespace NConcurrency;
+
+static NLogging::TLogger Logger("fio_plugin");
+
+//
+
+DEFINE_ENUM(EClientType,
+    (Http)
+    (Rpc)
+    (Native)
+);
+
+DECLARE_REFCOUNTED_CLASS(TPluginConfig)
+
+class TPluginConfig
+    : public TSingletonsConfig
+{
+public:
+    EClientType ClientType;
+    bool PrintFileChunks;
+
+    NApi::TFileReaderConfigPtr FileReaderConfig;
+    NApi::TFileWriterConfigPtr FileWriterConfig;
+
+    REGISTER_YSON_STRUCT(TPluginConfig);
+
+    static void Register(TRegistrar registrar) {
+        registrar.Parameter("client_type", &TThis::ClientType)
+            .Default(EClientType::Http);
+        registrar.Parameter("print_file_chunks", &TThis::PrintFileChunks)
+            .Default();
+        registrar.Parameter("file_reader_config", &TThis::FileReaderConfig)
+            .DefaultNew();
+        registrar.Parameter("file_writer_config", &TThis::FileWriterConfig)
+            .DefaultNew();
+    }
+};
+
+DEFINE_REFCOUNTED_TYPE(TPluginConfig);
+
+//
+
+DECLARE_REFCOUNTED_STRUCT(TFile)
+
+struct TFile
+    : public TRefCounted
+{
+    TFile(TStringBuf name)
+        : Name(name)
+    {
+    }
+
+    TString Name;
+
+    uint64_t ReaderOffset;
+    IFileReaderPtr Reader;
+    IFileWriterPtr Writer;
+
+    IAsyncInputStreamPtr AsyncReader;
+    NApi::IFileWriterPtr AsyncWriter;
+
+    TFuture<void> Future;
+
+    bool UnlinkAtClose = false;
+};
+
+DEFINE_REFCOUNTED_TYPE(TFile)
```

Comment:
It seems that we usually write `DECLARE_REFCOUNTED_STRUCT` and `DEFINE_REFCOUNTED_TYPE` side by side